python/src/trezorlib/tools.py

```diff
@@ -133,26 +160,26 @@ def b58decode(v, length=None):
             break
 
     result = b"\x00" * nPad + result
-    if length is not None and len(result) != length:
-        return None
+    if expected_length is not None and len(result) != expected_length:
+        raise ValueError("Result length does not match expected_length")
 
     return result
 
 
-def b58check_encode(v):
+def b58check_encode(v: bytes) -> str:
     checksum = btc_hash(v)[:4]
     return b58encode(v + checksum)
 
 
-def b58check_decode(v, length=None):
+def b58check_decode(v: Union[str, bytes], length: Optional[int] = None) -> bytes:
     dec = b58decode(v, length)
     data, checksum = dec[:-4], dec[-4:]
     if btc_hash(data)[:4] != checksum:
         raise ValueError("invalid checksum")
     return data
 
 
-def parse_path(nstr: str) -> Address:
+def parse_path(nstr: Optional[str]) -> Address:
```

Comment:
are you sure this is `Optional`? do we have usages where `nstr is None`?

Reply:
There is indeed no sign of using it without argument (just `parse_path()`). I put it there based on the `if not nstr:` check. However, it works fine even when removing the `Optional`, so I am getting rid of it. Maybe we could consider checking for empty string with `if nstr == "":` to further imply it is always a string

Reply:
Just found out about some "issues" in `cardano.py`, where we call `parse_path(address_parameters.get("path"))`, so it is thinking that it may receive `None`. I might change `get("path")` to `get("path", "")` - it would be cleaner to probably access it by `address_parameters["path"]`, but it could break somebody using this function with empty dict (as it would raise `KeyError`)

Reply:
Changes also needed in `cli/cardano.py`, where we supply arguments from `CLI` into `parse_path()`, so I have changed the default parameter from `None` to `""`.

Overall, not sure if this and the changes in comment above were worth removing the `Optional` type

Reply:
Done in 4cfe54a

Reply:
I'm not sure either but this way `parse_path()` seems more correct. Let's keep it this way for now.

Re Cardano, they're doing a thing where they're asking for all paths but only some of them are required. I'll review the particular usages, but I believe using `x.get("path", "")` and not raising is a deliberate choice.